MaxText/tokenizer.py

```diff
@@ -17,35 +17,94 @@
 """Provides op for tokenizing a dataset."""
 
 from typing import Any, Dict, Iterable
-
 import dataclasses
-from absl import logging
 import tensorflow as tf
 import tensorflow_text as tftxt
-
 import max_logging
+from jetstream.engine import tokenizer_api
+from jetstream.third_party.llama3 import llama3_tokenizer
 
 
 Features = Dict[str, tf.Tensor]
 
 
-def _load_sentencepiece_tokenizer(tokenizer_path: str, add_bos: bool = False, add_eos: bool = True, reverse: bool = False):
-  """Load a tf-text SentencePiece tokenizer from given model filepath."""
+class TikToken(tokenizer_api.Tokenizer):
+  """Tiktoken tokenizer"""
+  def __init__(self, tokenizer_path: str, add_bos: bool, add_eos: bool):
+    self.tokenizer = llama3_tokenizer.Tokenizer(tokenizer_path)
+    self.add_bos = add_bos
+    self.add_eos = add_eos
+
+  def encode(self, s):
+    t = []
+    t.extend(self.tokenizer.model.encode(str(s)))
+    if self.add_bos:
+      t.insert(0, self.tokenizer.bos_id)
+    if self.add_eos:
+      t.append(self.tokenizer.eos_id)
+    return t
+
+  def decode(self, token_ids):
+    return self.tokenizer.model.decode(token_ids)
+
+  def decode_str(self, token_ids):
+    # same as decode
+    return self.decode(token_ids)
+
+  @property
+  def pad_id(self) -> int:
+    """ID of the pad token."""
+    return self.tokenizer.pad_id
+
+  @property
+  def eos_id(self) -> int:
+    """ID of EOS token."""
+    return self.tokenizer.eos_id
+
+  @property
+  def bos_id(self) -> int:
+    """ID of the BOS token."""
+    return self.tokenizer.bos_id
+
+class SentencePieceTokenizer(tokenizer_api.Tokenizer):
+  """SentencePiece tokenizer"""
+  def __init__(self, tokenizer_path: str, add_bos: bool, add_eos: bool):
+    with tf.io.gfile.GFile(tokenizer_path, "rb") as model_fp:
+      sp_model = model_fp.read()
+    self.sp_tokenizer = tftxt.SentencepieceTokenizer(model=sp_model, add_bos=add_bos, add_eos=add_eos)
+
+  def encode(self, s):
+    return self.sp_tokenizer.tokenize(s)
+
+  def decode(self, token_ids):
+    return self.sp_tokenizer.detokenize(token_ids)
+
+  def decode_str(self, token_ids):
+    # same as decode
+    return self.decode(token_ids)
+
+  @property
+  def pad_id(self) -> int:
+    """ID of the pad token."""
+    return 0
+
+  @property
+  def eos_id(self) -> int:
+    """ID of EOS token."""
+    return self.sp_tokenizer.detokenize('</s>')
```

Comment:
the input to detokenize must be a list of ints and not a string?

Reply:
Changed it to tokenize. Thanks